sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -927,6 +937,179 @@ abstract class GpuBaseAggregateMeta[INPUT <: SparkPlan](
+      }
+      GpuHashAggregateExec(
+        requiredChildDistributionExpressions.map(_.map(_.convertToGpu())),
+        groupingExpressions.map(_.convertToGpu()),
+        aggregateExpressions.map(_.convertToGpu()).asInstanceOf[Seq[GpuAggregateExpression]],
+        aggAttributes.map(_.convertToGpu()).asInstanceOf[Seq[Attribute]],
+        retExpressions.map(_.convertToGpu()).asInstanceOf[Seq[NamedExpression]],
+        childPlans.head.convertIfNeeded(),
+        conf.gpuTargetBatchSizeBytes)
+    } else {
+      super.convertToGpu()
+    }
+  }
+
+  /**
+   * The method replaces data types of aggregation buffers created by TypedImperativeAggregate
+   * functions with the actual data types used in the GPU runtime. And this method can adapt all
+   * Aggregation modes, including PartialMerge.
+   *
+   * Firstly, this method traverses aggregateFunctions, to search attributes referring to
+   * aggregation buffers of TypedImperativeAggregate functions.
+   * Then, we extract the desired (actual) data types on GPU runtime for these attributes,
+   * and map them to expression IDs of attributes.
+   * At last, we traverse aggregateAttributes and resultExpressions, overriding data type in
+   * RapidsMeta if necessary, in order to ensure TypeChecks tagging exact data types in runtime.
+   */
+  private def overrideAggBufTypes(): Unit = {
+    val desiredAggBufTypes = mutable.HashMap.empty[ExprId, DataType]
+    val desiredInputAggBufTypes = mutable.HashMap.empty[ExprId, DataType]
+    // Collects exprId from TypedImperativeAggBufferAttributes, and maps them to the data type
+    // of `TypedImperativeAggExprMeta.aggBufferAttribute`.
+    agg.aggregateExpressions.zipWithIndex.foreach {
+      case (expr, i) if expr.aggregateFunction.isInstanceOf[TypedImperativeAggregate[_]] =>
+
+        val aggFn = expr.aggregateFunction
+        val aggMeta = aggregateExpressions(i).childExprs.head
+            .asInstanceOf[TypedImperativeAggExprMeta[_]]
+        val desiredDataType = aggMeta.aggBufferAttribute.dataType
+
+        var buf = aggFn.aggBufferAttributes.head
+        desiredAggBufTypes(buf.exprId) = desiredDataType
+
+        buf = aggFn.inputAggBufferAttributes.head
+        desiredInputAggBufTypes(buf.exprId) = desiredDataType
+
+      case _ =>
+    }
+
+    // Overrides the data types of typed imperative aggregation buffers for type checking
+    aggregateAttributes.foreach { attrMeta =>
+      attrMeta.wrapped match {
+        case ar: AttributeReference if desiredAggBufTypes.contains(ar.exprId) =>
+          attrMeta.overrideDataType(desiredAggBufTypes(ar.exprId))
+        case _ =>
+      }
+    }
+    resultExpressions.foreach { retMeta =>
+      retMeta.wrapped match {
+        case ar: AttributeReference if desiredInputAggBufTypes.contains(ar.exprId) =>
+          retMeta.overrideDataType(desiredInputAggBufTypes(ar.exprId))
+        case _ =>
+      }
+    }
+  }
+}
+
+object GpuNoHashAggregateMeta {
+  private val logicalIdToMetas =
```

Comment:
Is this going to continue to grow as the application runs? It seems like so because this is a static class.

I think the main reason for this is so that a half aggregate isn't currently doable for the typed imperative case. Is that right? Did you try other ways of handling it?

Reply:
Yes, here is a potential risk. And I reworked this part with a different approach.